docs/csharp/pattern-matching.md

```diff
@@ -2,31 +2,192 @@
+
+```csharp
+Square s;
+if (shape is Square)
+{
+    s = (shape as Square);
+    // use 's'
+} 
+```
+
+Language rules for pattern matching expressions help you avoid misuising
+the results of a match expression. In the example above, the variables `s`
+and `c` are only in scope and definitely assigned when the respective
+pattern match expressions have true results. If you try and use either
+variable in another location, your code generates compiler errors.
+
+Let's examine both of those rules in detail, beginning with scope. The variable
+`c` is in scope only in the `else` branch of the first `if` statement. The variable
+`s` is in scope in the method `ComputeArea`. That's because the true and false
+branches of an `if` statement each establish a scope. However, the `if` statement
+itself does not. That means variables declared in the `if` statement are in the
+same scope as the `if` statement (the method in this case.) 
+
+The variables `c` and `s` are assigned when the respective `if` statements are true
+because of the definitely assigned when true mechanism. (You could reverse the
+logic by saying `if !(shape is Square s)` and the variable `s` would be definitely
+assigned only in the false branch.) 
+
+These rules mean that you are unlikely to accidentally access the result
+of a pattern match expression when that pattern ws not met.
+
+## Extending to more shape types
+
+As time goes on, you need to support other shape types. As the number
+of conditions you are testing grows, you'll find that using the `is` pattern
+matching expressions becomes cumbersome. In addition to requiring `if`
+statements on each type you want to check, the `is` expressions are limited
+to testing a type.
+
+As the number of conditions grows, you'll find that the `switch` pattern
+matching expressions becomes a better choice. To get ready to add more
+types of shapes to this library, let's change the existing method to use
+the `switch` pattern match expression:
+
+[!code-csharp[ComputeWithSwitchExpression](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#06_ComputeWithSwitchExpression "Compute with switch expression")]
+
+The pattern matching `switch` statement uses familiar syntax to developers
+that have used the traditional `switch` statement. Each `case` is evaluated
+and the code beneath the condition that matches the input variable is
+executed. Code execution cannot "fall through" from one `case` expression
+to the next. Each block following a `case` must end with a `break`, `return`
+or `goto`.
+
+There are important new rules governing the `switch` statement. The restrictions
+on the type of the variable in the `switch` expression have been removed.
+Any type, such as `object` in this example, may be used. The case expressions
+are no longer limited to constant values. Removing that limitation means
+that reordering switch sections may change a program's behavior.
+
+When limited to constant values, no more than one switch
```

Comment:
I found this a bit difficult to parse.  Perhaps it's better to go through the rules for how things are evaluated in a numerical, step-by-step fashion.  You could then follow that up with examples of only constants, only reference types, and mixed/constant reference types.  It may be worth expounding on how the ordering of the cases could affect behavior, and finish up with general guidance.